akka-http-core/src/main/scala/akka/http/impl/engine/ws/Handshake.scala

```diff
@@ -62,50 +64,67 @@ private[http] object Handshake {
      *        to speak.  The interpretation of this header field is discussed
      *        in Section 9.1.
      */
-    def websocketUpgrade(headers: List[HttpHeader], hostHeaderPresent: Boolean): Option[UpgradeToWebSocket] = {
-      def find[T <: HttpHeader: ClassTag]: Option[T] =
-        headers.collectFirst {
-          case t: T ⇒ t
-        }
+    def websocketUpgrade(headers: List[HttpHeader], hostHeaderPresent: Boolean): OptionVal[UpgradeToWebSocket] = {
 
       // Host header is validated in general HTTP logic
-      // val host = find[Host]
-      val upgrade = find[Upgrade]
-      val connection = find[Connection]
-      val key = find[`Sec-WebSocket-Key`]
-      val version = find[`Sec-WebSocket-Version`]
+      // val host = HttpHeader.fastFind[Host](headers)
+
       // Origin header is optional and, if required, should be validated
       // on higher levels (routing, application logic)
-      // val origin = find[Origin]
-      val protocol = find[`Sec-WebSocket-Protocol`]
-      val clientSupportedSubprotocols = protocol.toList.flatMap(_.protocols)
+      // val origin = HttpHeader.fastFind[Origin](headers)
+
       // Extension support is optional in WS and currently unsupported.
       // TODO See #18709
-      // val extensions = find[`Sec-WebSocket-Extensions`]
-
-      if (upgrade.exists(_.hasWebSocket) &&
-        connection.exists(_.hasUpgrade) &&
-        version.exists(_.hasVersion(CurrentWebSocketVersion)) &&
-        key.exists(k ⇒ k.isValid)) {
+      // val extensions = HttpHeader.findFast[`Sec-WebSocket-Extensions`](headers)
+
+      // these are not needed directly, we verify their presence and correctness only:
+      // val upgrade = HttpHeader.fastFind[Upgrade](headers)
+      // val connection = HttpHeader.fastFind[Connection](headers)
+      // val version = HttpHeader.fastFind[`Sec-WebSocket-Version`](headers)
+      def hasAllRequiredWebsocketUpgradeHeaders: Boolean = {
+        // single-pass through the headers list while collecting all needed requirements
+        // this way we avoid scanning the requirements list 3 times (as we would with collect/find)
+        val it = headers.iterator
+        var requirementsMet = 0
+        val targetRequirements = 3
+        while (it.hasNext && (requirementsMet != targetRequirements)) it.next() match {
+          case u: Upgrade                 ⇒ if (u.hasWebSocket) requirementsMet += 1
+          case c: Connection              ⇒ if (c.hasUpgrade) requirementsMet += 1
+          case v: `Sec-WebSocket-Version` ⇒ if (v.hasVersion(CurrentWebSocketVersion)) requirementsMet += 1
+          case _                          ⇒ // continue...
+        }
+        requirementsMet == targetRequirements
+      }
```

Comment:
Ah, nice, just counting them. One idea, maybe for later, I guess one could do an even more imperative loop and keep the index at which each was find to avoid the find-part for when it actually is a websocket request.